# 2021-07-21 파이썬 공부

## **< 오전 >**

### 함수

특정한 기능(function)을 하는 코드의 묶음(일의 단위)

- **함수 기초**

import statistics

statistics.pstdev() : 표준편차, statistics.mean() : 평균

Jupyter notebook에서 함수에 커서를 놓고 shift + tap 누르면 Docstring 볼 수 있음

또는 pstdev.\_\_doc\_\_

내장함수(Built-in Functions)

Docstring은 일반적으로 큰 따옴표 3개로 작성



- **함수의 Output**

함수는 동작 후에 return을 통해 결과 값을 전달함, 반드시 **하나**의 객체를 반환

```python
# 복수의 객체를 return 하는 경우
a, b = 3, 2
def foo(a, b):
    return a+b, a-b
print(foo(a, b))
```

```
(5, 1) # 결국 하나의 tuple 객체를 반환함
```



```python
# 명시적인 return 값이 없는 경우
def greeting():
    print('hi') # 반환 값이 없는데요?
greeting()
```

```
hi
```

```python
type(greeting())
```

```
NoneType
```

```python
my_var = greeting() # my_var은 하나의 객체(None)
print(my_var)
```

```
None
```



- **함수의 Input**

함수에서의 **매개변수(parameter)** vs **인자(전달인자, 인수)(argument)**

**매개변수**는 함수가 선언될 때 입력으로 전달된 값을 받는 변수(x, y)

**인자**는 함수를 호출할 때 함수에 전달하는 입력 값(2, 3)

```python
def add(x, y):
    # x = 2; y = 3
    return x + y
add(2, 3) # 주석처럼 들어감, 파이썬은 call by assignment(reference, value X)
```

```python
def add(x, y=0): # 여기서 y는 기본인자
    # x = 2
    return x + y
add(2), add(2, 3), add(2, y=3) # 키워드 인자
# 키워드 인자 다음에 위치 인자를 활용할 수 없음
```

```
(2, 5, 5)
```




- **가변 인자**

```python
# 가변 인자 리스트(*), 인자들은 튜플로 묶여 처리(packing하여 처리)
def add(*args): # args의 이름은 바꿔도 되지만, 관례적으로 args를 씀
    total = 0
    for arg in args:
        total += arg
    return total
add(2), add(2, 3, 4, 5)
```

```
(2, 14)
```



```python
# 가변 키워드 인자(**), 인자들은 딕셔너리로 묶여 처리
def family(**kwargs):
    for key, value in kwargs.items():
        print(key, ":", value)
family(father='John', mother='Jane', me='John Jr.')
```

```
father : John
mother : Jane
me : John Jr.
```



```python
# 가변 인자리스트, 가변 키워드 인자는 위치 인자보다 앞쪽에 올 수 없음
def add(*args, x):
    pass
add(1, 2, 3) # x가 뭔지 알 수 없기 때문에 사용할 수 없다!
```



```python
# 올바른 순서
def my_info(x, y, *args, **kwargs):
    pass
# 위치 인자, 기본값이 있는 위치 인자, 가변 인자, 가변 키워드 인자 순서로 작성하는 것이 가장 좋다.
```



```python
def func(a, b=1, *args, **kwargs):
    return a, b, args, kwargs
print(func(2))
print(func(2, 3))
print(func(2, 10, 20, 30, father='John', mother='Jane'))
# *args랑 **kwargs의 순서가 바뀌면 SyntaxError
```

```
(2, 1, (), {})
(2, 3, (), {})
(2, 10, (20, 30), {'father': 'John', 'mother': 'Jane'})
```



```python
a, *b, c = range(5)
print(a, b, c) # 함수 매개변수로 사용되는 경우가 아닐 때는 이렇게 적용되네
```

```
0 [1, 2, 3] 4
```

기본 인자 값을 가지는 인자 다음에 기본 값이 없는 인자로 정의할 수 없음



- **함수 스코프(scope)**

  - 전역 스코프 : 코드 어디에서든 참조할 수 있는 공간

  - 지역 스코프 : 함수가 만든 스코프. 함수 내부에서만 참조 가능

    ```python
    def func():
        pass # 이 공간이 지역 스코프
    ```

  - 전역 변수 : 전역 스코프에 정의된 변수

  - 지역 변수 : 지역 스코프에 정의된 변수

  ```python
  def func():
      a = 20
      print('local', a)
  func()
  print('global', a) # a는 함수 내에서 할당 되었으므로 함수가 종료되며 사라짐
  ```

  ```
  local 20
  NameError # name 'a' is not defined
  ```



- **변수 수명주기(lifecycle)**

  - 빌트인 스코프(built-in scope)

    파이썬이 실행된 이후부터 영원히 유지

  - 전역 스코프(global scope)

    모듈이 호출된 시점 이후 혹은 인터프리터가 끝날 때까지 유지

  - 지역(함수) 스코프(local scope)

    함수가 호출될 때 생성되고, 함수가 종료될 때까지 유지
    
    

- **이름 검색 규칙(Name Resolution) - LEGB Rule**

  Local scope : 함수

  Enclosed scope : 특정 함수의 상위 함수

  Global scope : 함수 밖의 변수, Import 모듈

  Built-in scope : 파이썬 안에 내장되어 있는 함수 또는 속성

  ```python
  dict = 'hi'
  dict() # 글로벌 스코프에서 dict가 정의되었기 때문에 내장 함수를 호출할 수 없음
  ```

  ```python
  # LEGB 연습
  a = 0
  b = 1
  def enclosed():
      a = 10
      c = 3
      def local(c):
          print(a, b, c)
      local(300)
  enclosed()
  print(a, b)
  ```

  ```
  10 1 300
  10 1 3
  0 1
  ```

함수 내에서 바깥 스코프의 변수에 접근은 가능하지만 (일반적으로) 수정할 수는 없음

**global** (변수명)을 활용하면 local scope 안에서 global scope의 변수 수정 가능, but 주의해서 사용해야 함(유지&보수, 일관성 측면)

**nonlocal**은 전역을 제외하고 가장 가까운(둘러 싸고 있는) 스코프의 변수를 연결하도록 함

global과는 달리 이미 존재하는 이름과의 연결만 가능함, global과 마찬가지로 주의해서 사용

global은 변수가 없다면 선언까지 가능하지만, nonlocal은 선언된 적 있는 변수에서만 사용 가능

**상위 스코프 변수가 필요하다면 인자로 받아서 활용하자!(클로저* 제외)**

(클로저 : 어떤 함수 내부에 중첩된 형태로써, 외부 스코프 변수에 접근 가능한 함수)



- **궁금한 게 생겼을 때 찾아봐야 할 순서**
1. 공식 문서 https://docs.python.org/3/
  
2. stackoverflow https://www.stackoverflow.com
  
3. 블로그(자료 검증X) - 안본다고 생각하는게 편함



## < 오후 >

```python
# 1
def cube(num):
    return num ** 3
# 2
cube = lambda num: num ** 3
# 1과 2는 같다.
```

파이썬에서 함수는 값이구나, 변수에 담겨있구나 정도

random.sample() 함수의 내부

> https://github.com/python/cpython/blob/master/Lib/random.py#L385

```python
return (num1 if num1 > num2 else num2) # 괄호는 1+1과 같이 값이 존재하는 식이다.
```

개발할 때 print()는 나를 믿지 못해서 디버깅할 때만 사용, 그리고 알고리즘 때만 print() 사용

```
# 함수 매개변수(parameter)/인자(argument) 관련
정의할 때 parameter 순서
basic -> default
호출할 때 argument
position -> keyword
```

**asterisk** : 별표(*), **caret** : (^)



- **재귀 함수**

알고리즘 중 재귀 함수로 로직을 표현하기 쉬운 경우(ex. 점화식 등)

1개 이상의 base case(종료되는 상황)가 존재하고, 수렴해야 함

큰 문제를 해결하기 위해 작은 문제로 좁히고, 작은 문제의 해답을 이용하여 해결(ex. 팩토리얼, 피보나치 등)

재귀에서 메모리 스택이 넘치게 되면(stack overflow) 프로그램이 동작하지 않음

파이썬에서는 최대 재귀 깊이(maximum recursion depth) == 1000, 호출 횟수가 이를 넘어가면 RecursionError 발생



- **에러/예외 처리(Error/Exception Handling)**

  - 디버깅

    print 문 활용

    개발 환경 등에서 제공하는 기능 활용

    Python tutor 활용

    뇌컴파일, 눈디버깅

EOL(End of Line), EOF(End of File)

- **예외(Exception)**

  - 파이썬 내장 예외(built-in-exceptions)

    *SyntexError를 제외한 나머지들*

    ZeroDivisionError : 0으로 나누고자 할 때 발생

    NameError : namespace 상에 이름이 없는 경우

    TypeError : 타입 불일치, argument(인자) 누락, argument 개수 초과, argument type 불일치

    ValueError : 타입은 올바르나 값이 적절하지 않거나 없는 경우

    IndexError : 인덱스가 존재하지 않거나 범위를 벗어나는 경우

    KeyError : 해당 키가 존재하지 않는 경우

    ModuleNotFoundError : 존재하지 않는 모듈을 import 하는 경우

    ImportError : Module은 있으나 존재하지 않는 클래스/함수를 가져오는 경우

    KeyboardInterrupt : 임의로 프로그램을 종료하였을 때

    IndentationError : Indentation이 적절하지 않는 경우(Tab을 잘못 사용했을 때)

    ...

- **예외 처리**

```python
try:
    num = input('100을 나눌 값을 입력하시오: ')
    100/int(num)
# except (ValueError, ZeroDivisionError):
#     print('제대로 입력해줘')
except Exception:
    print("에러는 모르지만 에러가 발생하였습니다.")
except ValueError: # 위에 범주가 아래를 포함해서 이 구문은 만날 수 없다
    print('숫자를 넣어줘')
# except는 순차적으로 수행되기 때문에 가장 작은 범주부터 예외 처리해야 함
```

파일을 열고 읽는 코드를 작성하는 경우

```python
try:
    f = open('file.txt') # 파일 열기
```

try : 코드를 실행함 (파일 열기 시도)

except : try 문에서 예외가 발생시 실행함 (파일 없는 경우)

else : try 문에서 예외가 발생하지 않으면 실행함 (파일 있는 경우)

finally : 예외 발생 여부와 관계없이 항상 실행함 (해당 파일 읽기 작업 종료 메시지 출력)



```python
try:
    empty_list = []
except IndexError as err: # as 키워드로 원본 에러 메시지 사용할 수 있음
    print(f'{err} 오류가 발생했습니다.)
```

```
list index out of range 오류가 발생했습니다.
```



- **예외 발생 시키기**

주로 디버깅할 때 사용

- **raise**

```python
raise ValueError('값 에러 발생')
```

```
ValueError: 값 에러 발생
```


  - **assert**


```python
assert len([1, 2]) == 1, '길이가 1이 아닙니다.'
```

```
AssertionError: 길이가 1이 아닙니다.
```

-O 옵션으로 실행하는 경우, assert문과 \_\_debug\_\_에 따른 조건부 코드를 제거 후 실행

(ex. python -O assert_example.py)

- **try-except와 if-else**

```python
# try문으로 바꿀 수 있는 if문이 있음
# 파이썬 공식 문서 glossary(용어집)에서는 try-except문을 권장함

my_dict = {'key': 'value'}
# EAFP, 허락보다 용서를 구하는 것이 쉽다, 예외처리를 활용하여 검사를 수행하지 않고 일단 실행하고 예외처리를 진행하는 스타일, 파이썬 코드가 문제 없이 '실행될 것'을 전제로 코드를 실행
try: 
    x = my_dict['key'] # key를 한번 검색
except KeyError:
    pass

# LBYL, 도약하기 전에 먼저 봐라, 어떤 것이 실행하기 전에 에러가 날만한 요소들을 조건문으로 검사를 하고 수행
if 'key' in my_dict: # key를 한번 검색
    x = my_dict['key'] # key를 한번 더 검색
else:
    pass
```



```python
d1 = {'a': 1, 'b': 2}
d2 = dict(a=1, b=2)
# d1과 d2는 같다
```

```python
# Key가 숫자인 딕셔너리를 만들고 싶다면, 아래와 같이 사용해야합니다.
dict([(1, 1), (2, 2)])
dict(((1,1), (2,2)))
# 또는 {1: 1, 2: 2}로 만들어도 됨
```



- 함수에서 리스트 값 바꾸면 함수 밖의 리스트 값이 바뀐다!(pythontutor 참고)

  -> 함수에서 리스트 안의 값을 바꾸지 않으려고 해야한다. 아니면 튜플 써라.

  -> 해야만 한다면 Copy를 떠야한다. - 자료구조에서 나옴

```python
def change_list(numbers):
    numbers[0] = 100
    return numbers

my_list = [1, 2, 3, 4, 5]
change_list(my_list), my_list
```

```
([100, 2, 3, 4, 5], [100, 2, 3, 4, 5])
```



- 리스트, 딕셔너리같은 경우에 trailing comma가 권장된다.(다음에 들어올 데이터를 위해서)



- reduce 함수

```python
from functools import reduce
reduce(lambda n1, n2: n1 + n2, [1, 2, 3, 4, 5])
```

```
15
```



